File: liftie-bot.py
import discord
import json
import os
from discord import app_commands
from dotenv import load_dotenv

# Load environment variables
load_dotenv()
TOKEN = os.getenv("BOT_TOKEN")

# Load resorts from JSON
lifts_file = 'lifts_status.json'
with open(lifts_file, 'r') as file:
    resorts_data = json.load(file)
ALL_RESORTS = [resort['resort'] for resort in resorts_data]  # Extract resort names

# Set up bot with commands
intents = discord.Intents.default()
client = discord.Client(intents=intents)
tree = app_commands.CommandTree(client)


# Auto-complete function (filters resorts dynamically)
async def resort_autocomplete(interaction: discord.Interaction, current: str):
    return [
               app_commands.Choice(name=resort, value=resort)
               for resort in ALL_RESORTS if current.lower() in resort.lower()
           ][:25]  # Limit to 25 options


# Slash command for lift status
@tree.command(name="lifts", description="Get lift status for a resort")
@app_commands.autocomplete(resort=resort_autocomplete)
async def lifts(interaction: discord.Interaction, resort: str):
    resort_data = next((r for r in resorts_data if r['resort'].lower() == resort.lower()), None)

    if not resort_data:
        await interaction.response.send_message(f"Resort '{resort}' not found.", ephemeral=True)
        return

    lifts_info = resort_data['lifts']
    response = f"**Lift status for {resort}:**\n" + "\n".join(
        f"{lift['name']}: {'✅' if lift['status'].lower() == 'open' else '❌'} {lift['status']}"
        for lift in lifts_info
    )

    await interaction.response.send_message(response)


# Event when bot is ready
@client.event
async def on_ready():
    await tree.sync()
    print(f"Bot is online as {client.user}")


# Run bot
client.run(TOKEN)

File: liftie-bot.js
const { Client, Intents, SlashCommandBuilder } = require('discord.js');
const fs = require('fs');
const path = require('path');
require('dotenv').config();

const client = new Client({
  intents: [
    Intents.FLAGS.GUILDS,
    Intents.FLAGS.GUILD_MESSAGES,
    Intents.FLAGS.DIRECT_MESSAGES,
  ],
});

const liftsFile = 'lifts_status.json'; // Path to your lifts status JSON file

// Function to read the lift status from the JSON file
function getLiftStatus(resortName) {
  try {
    const data = JSON.parse(fs.readFileSync(liftsFile, 'utf8'));
    const resortData = data.find(
      (resort) => resort.resort.toLowerCase() === resortName.toLowerCase()
    );

    if (!resortData) {
      return `Resort '${resortName}' not found in the data. Please make sure the resort name is correct.`;
    }

    const liftsInfo = resortData.lifts;
    let liftStatusMsg = `**Lift status for ${resortName.charAt(0).toUpperCase() + resortName.slice(1)} Resort:**\n`;

    liftsInfo.forEach((lift) => {
      const statusEmoji = lift.status === 'Closed' ? '❌' : '';
      liftStatusMsg += `${lift.name}: ${statusEmoji} ${lift.status}\n`;
    });

    return liftStatusMsg;
  } catch (error) {
    console.error(error);
    return 'Error reading the lift status file.';
  }
}

// Function to get available resort names
function getAvailableResorts() {
  try {
    const data = JSON.parse(fs.readFileSync(liftsFile, 'utf8'));
    const resortNames = data.map((resort) => resort.resort);
    return resortNames;
  } catch (error) {
    console.error(error);
    return 'Error fetching resort names.';
  }
}

// Slash commands registration
client.once('ready', async () => {
  const guildId = 'YOUR_GUILD_ID'; // Replace with your server's Guild ID
  const guild = await client.guilds.fetch(guildId);
  const commands = guild.commands;

  // Register slash commands
  await commands.set([
    new SlashCommandBuilder()
      .setName('lifts')
      .setDescription('Get the lift status for a resort')
      .addStringOption((option) =>
        option
          .setName('resort')
          .setDescription('The name of the resort')
          .setRequired(true)
          .setAutocomplete(true)
      ),
    new SlashCommandBuilder()
      .setName('lifts_resorts')
      .setDescription('List all available resorts'),
  ]);

  console.log('Bot is online!');
});

// Auto-completion for the resort name argument
client.on('interactionCreate', async (interaction) => {
  if (!interaction.isCommand()) return;

  const { commandName, options } = interaction;

  if (commandName === 'lifts') {
    const resortName = options.getString('resort');
    const liftStatus = getLiftStatus(resortName);
    await interaction.reply(liftStatus);
  } else if (commandName === 'lifts_resorts') {
    const resorts = getAvailableResorts();
    await interaction.reply(resorts.join(', '));
  }
});

// Auto-completion logic
client.on('interactionCreate', async (interaction) => {
  if (interaction.isAutocomplete()) {
    const focusedOption = interaction.options.getFocused();
    const availableResorts = getAvailableResorts();
    const filteredResorts = availableResorts.filter((resort) =>
      resort.toLowerCase().startsWith(focusedOption.toLowerCase())
    );

    await interaction.respond(
      filteredResorts.map((resort) => ({ name: resort, value: resort }))
    );
  }
});

// Log in to Discord with your app's token
client.login(process.env.BOT_TOKEN);
